pykokkos/interface/type_inference.py

```diff
@@ -0,0 +1,234 @@
+    '''
+
+    param_list = list(inspect.signature(handled_args.workunit).parameters.values())
+    args_list = list(*args)
+    updated_types = UpdatedTypes(workunit=handled_args.workunit, inferred_types={}, is_arg=set(), layout_change={})
+    policy_params: int = len(handled_args.policy.begin) if isinstance(handled_args.policy, MDRangePolicy) else 1
+
+    # accumulator 
+    if parallel_type == "parallel_reduce":
+        policy_params += 1
+    # accumulator + lass_pass
+    if parallel_type == "parallel_scan":
+        policy_params += 2
+
+    # Handling policy parameters
+    for i in range(policy_params):
+        param = param_list[i]
+
+        if param.annotation is inspect._empty:
+            # Check policy and apply annotation(s)
+            if isinstance(handled_args.policy, RangePolicy) or isinstance(handled_args.policy, TeamThreadRange):
+                # only expects one param
+                if i == 0:
+                    updated_types.inferred_types[param.name] = "int"
+                    updated_types.is_arg.add(param.name)
+
+            elif isinstance(handled_args.policy, TeamPolicy):
+                if i == 0:
+                    updated_types.inferred_types[param.name] = 'pk.TeamMember'
+                    updated_types.is_arg.add(param.name)
+
+            elif isinstance(handled_args.policy, MDRangePolicy):
+                total_dims = len(handled_args.policy.begin) 
+                if i < total_dims:
+                    updated_types.inferred_types[param.name] = "int"
+                    updated_types.is_arg.add(param.name)
+            else:
+                raise ValueError("Automatic annotations not supported for this policy")
+
+            # last policy param for parallel reduce and second last for parallel_scan is always the accumulator; the default type is double
+            if i == policy_params - 1 and parallel_type == "parallel_reduce" or i == policy_params - 2 and parallel_type == "parallel_scan":
+                updated_types.inferred_types[param.name] = "Acc:double"
+                updated_types.is_arg.add(param.name)
+
+            if i == policy_params - 1 and parallel_type == "parallel_scan":
+                updated_types.inferred_types[param.name] = "bool"
+                updated_types.is_arg.add(param.name)
+
+    if len(param_list) == policy_params:
+        if not len(updated_types.inferred_types): return None
+        return updated_types
+
+    # Handle Keyword args, make sure they are treated by queuing them in args
+    if len(passed_kwargs.keys()):
+        # add value to arguments list so the value can be assessed
+        for param in param_list[policy_params:]:
+            if param.name in passed_kwargs:
+                args_list.append(passed_kwargs[param.name])
+
+    # Handling arguments other than policy args, they begin at value_idx in args list
+    value_idx: int = 3 if handled_args.name != None else 2 
```

Comment:
It's not clear to me why this is 3 or 2

Reply:
updating comments to:
`  # Handling arguments other than policy args, they begin at value_idx in args list
   # e.g idx=3 -> parallel_for("label", policy, workunit, <other args>...) or if "label" is missing: 2 `

Reply:
Please resolve this if you think this should be good